token-cli: Support the `InterestBearingMint` extension

udbhav1 opened the pull request:
resolves #3406 

i used `handle_tx` for `command_set_interest_rate` because i was getting a trait bound error trying to pass an `Arc<dyn Signer>` into the token client's `update_interest_rate`

not super happy with how i handled the two params for `--interest-bearing <RATE_AUTHORITY> <RATE_BPS>` - i made the former optional by checking if it was "NONE" because i couldn't find a way to do this natively in clap

2501babe commented:
nice work! ill be able to look at this in more detail later

as for the token client, ive been in the process of reworking it to implement `close-mint`. you can see in full the approach ive been taking in #3487, but the tldr is we probably want to change the client interface to look like this:
```
    /// Update interest rate
    pub async fn update_interest_rate<S: Signers>(
        &self,
        authority: &Pubkey,
        new_rate: i16,
        signing_keypairs: &S,
    ) -> TokenResult<T::Output> {
```

2501babe commented on token/cli/src/main.rs:
```diff
@@ -1961,9 +2026,45 @@ fn app<'a, 'b>(
                             "Enable the mint authority to freeze associated token accounts."
                         ),
                 )
+                .arg(
+                    Arg::with_name("interest_bearing")
+                        .long("interest-bearing")
+                        .value_names(&["ADDRESS", "RATE"])
+                        .help("Specify the interest rate authority (or NONE) and rate in basis points"),
+                )
```
youre right that this is a tricky interface. i dont like the double parameter, it feels confusing and un-unixy. im thinking the most reasonable thing to do is make the flag `--interest-rate`, accept just the rate, and make the authority same as mint authority (as is done with `--enable-freeze` and `--enable-close` already). users can use `set-authority` to change this (it turns out `command_authorize` doesnt support all the new types yet, but ill open a separate issue, dont worry about this for your own pr)

2501babe commented on token/cli/src/main.rs:
```diff
@@ -2702,6 +2803,23 @@ async fn process_command<'a>(
                 config.pubkey_or_default(arg_matches, "mint_authority", &mut wallet_manager);
             let memo = value_t!(arg_matches, "memo", String).ok();
 
+            let mut rate_authority: Option<Pubkey> = None;
+            let mut rate_bps: Option<i16> = None;
+            let interest_bearing =
+                values_t!(arg_matches.values_of("interest_bearing"), String).ok();
+            if let Some(values) = interest_bearing {
+                let rate_addr = &values[0];
+                if rate_addr != "NONE" {
+                    is_valid_pubkey(rate_addr)?;
+                    rate_authority = Some(config.pubkey_or_default(
+                        arg_matches,
+                        "interest_bearing",
+                        &mut wallet_manager,
+                    ));
+                }
+                rate_bps = Some(values[1].parse::<i16>().unwrap());
+            }
+
```
this becomes a much more pleasant `values_t!` call if we change the flag to one param

2501babe commented on token/cli/src/main.rs:
```diff
@@ -385,6 +397,59 @@ async fn command_create_token(
+        ),
+    );
+
+    let minimum_balance_for_rent_exemption = if !config.sign_only {
+        config
+            .program_client
+            .get_minimum_balance_for_rent_exemption(Account::LEN)
+            .await?
+    } else {
+        0
+    };
```
this is unnecessary, no account is being created so the arg passed to `handle_tx` can always be zero. however we should do basic sanity checks inside the `if !config.sign_only` block. off the top of my head i would... fetch mint, check that its owned by token22, check that it has the interest extension, check that the rate authority is correct. you dont need to write this, i can add it after if you prefer

2501babe commented on token/cli/src/main.rs:
```diff
@@ -356,13 +359,22 @@ async fn command_create_token(
     }
 
     let freeze_authority = if enable_freeze { Some(authority) } else { None };
+    let mut extensions = vec![];
+    if let Some(rate_bps) = rate_bps {
+        extensions.push(
+            spl_token_client::token::ExtensionInitializationParams::InterestBearingConfig {
+                rate_authority,
+                rate: rate_bps,
+            },
+        )
+    }
```
youll have to rebase against master because we create this vector now (as part of the close-authority pr), but you shouldnt have to do anything different otherwise

2501babe commented:
good work! i think if you could rebase against master and change the flag as described, this will be good to merge. i can handle the sanity checks on the mint if youd prefer (if so do you mind if i push to your branch?). im still in the process of converting client interfaces so we can use `handle_tx` for now

udbhav1 commented:
fixed the flag and hopefully i rebased properly. i think i'd rather have you do the mint sanity checks and i don't mind if you use my branch. i left a `// TODO` at the appropriate location. thanks for the review!

2501babe commented:
thanks for your contribution!
